api/src/game_handlers.rs

```diff
@@ -0,0 +1,655 @@
+            Err(_) => {
+                let msg = format!("unable to parse {} into usize.  Bad CSV.", token);
+                return HttpResponse::BadRequest().body(
+                    serde_json::to_string(&CribbageError::new(CribbageErrorKind::BadCount, msg))
+                        .unwrap(),
+                );
+            }
+        };
+    }
+    indices.push(shared_card_index);
+    let response = match get_random_hand_internal(is_computer_crib, indices) {
+        Some(response) => response,
+        None => {
+            return HttpResponse::BadRequest().body(
+                serde_json::to_string(&CribbageError::new(
+                    CribbageErrorKind::BadCount,
+                    "Error getting random hand".to_string(),
+                ))
+                .unwrap(),
+            );
+        }
+    };
+
+    HttpResponse::Ok().body(serde_json::to_string(&response).unwrap())
+}
+/**
+ *  Tests for the Web API.  The actual logic of the game is already tested in the unit tests for that part of the project
+ *
+ *  these tests test the "shape" of the Web API - making sure that serialization/deserialization works (by simply using it
+ *  in the tests) and that the response we get back are non-empty or have other reasonable values.
+ *
+ *  reproducibility isn't something the game uses, but is verified here to enforce the correctness of the repeat URL
+ *
+ */
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::game_handlers;
+    use actix_web::{test, web, App};
+    use cribbage_library::scoring::CombinationName;
+
+    macro_rules! get_repeat_url {
+        ($url: expr) => {{
+            let index = $url.find("/").unwrap() + 1;
+            let repeat_url = $url
+                .chars()
+                .skip(index - 1) // we want the '/' to start the string
+                .take($url.len() - index + 1)
+                .collect::<String>();
+
+            println!("repeat url: {}", repeat_url);
+            repeat_url
+        }};
+    }
+
+    macro_rules! vec_contains_card {
+        ($vec: expr, $card: expr) => {{
+            let mut found = false;
+            for c in $vec.iter() {
+                if c.cardName == $card {
+                    found = true;
+                    break;
+                }
+            }
+            found
+        }};
+    }
+    #[allow(unused_macros)]
+    macro_rules! vec_contains_score {
+        ($vec: expr, $name: expr) => {{
+            let mut found = false;
+            for score in $vec.iter() {
+                if score.ScoreName == $name {
+                    found = true;
+                    break;
+                }
+            }
+            found
+        }};
+    }
+
+    /**
+     *  this macro takes in the service to call, the orignal response, and the repeat URL
+     *  it then calls the service again with the repeat URL and verifies that the return
+     *  values are identical.
+     */
+    macro_rules! test_repeatability {
```

Comment:
Is this repeatability or idempotency?

Reply:
I guess both! :P  I use it in the client when random numbers are generated so that I can test it again.  especially valuable when you have a hand and you want to get a score and you see there is a scoring bug.  it lets you get back to the get_score call with the right hand.  and so on.